client/core/bot.go

```diff
@@ -117,77 +93,38 @@ func validateProgram(pgm *MakerProgram) error {
-		}
-	}
 
-	if pgm.DriftTolerance == 0 {
-		pgm.DriftTolerance = 0.001
+	if pgm.GapEngineCfg == nil && pgm.ArbEngineCfg == nil {
+		return fmt.Errorf("at least one engine cfg must be populated")
 	}
-	if pgm.DriftTolerance < 0 || pgm.DriftTolerance > 0.01 {
-		return fmt.Errorf("drift tolerance %f out of bounds", pgm.DriftTolerance)
+	if pgm.GapEngineCfg != nil && pgm.ArbEngineCfg != nil {
+		return fmt.Errorf("only one engine cfg can be populated")
 	}
 
-	var limits [2]float64
-	switch pgm.GapStrategy {
-	case GapStrategyMultiplier:
-		limits = [2]float64{1, 100}
-	case GapStrategyPercent, GapStrategyPercentPlus:
-		limits = [2]float64{0, 0.1}
-	case GapStrategyAbsolute, GapStrategyAbsolutePlus:
-		limits = [2]float64{0, math.MaxFloat64} // validate at < spot price at creation time
-	default:
-		return fmt.Errorf("unknown gap strategy %q", pgm.GapStrategy)
+	if pgm.GapEngineCfg != nil {
+		if err := pgm.GapEngineCfg.validate(); err != nil {
+			return fmt.Errorf("error validating gap engine cfg: %w", err)
+		}
 	}
 
-	if pgm.GapFactor < limits[0] || pgm.GapFactor > limits[1] {
-		return fmt.Errorf("%s gap factor %f is out of bounds %+v", pgm.GapStrategy, pgm.GapFactor, limits)
+	if pgm.ArbEngineCfg != nil {
+		if err := pgm.ArbEngineCfg.validate(); err != nil {
+			return fmt.Errorf("error validating arb engine cfg: %w", err)
+		}
 	}
-	return nil
-}
 
-// oracleWeighting returns the specified OracleWeighting, or the default if
-// not set.
-func (pgm *MakerProgram) oracleWeighting() float64 {
-	if pgm.OracleWeighting == nil {
-		return defaultOracleWeighting
-	}
-	return *pgm.OracleWeighting
+	return nil
 }
 
 // makerAsset combines a *dex.Asset with a WalletState.
 type makerAsset struct {
 	*SupportedAsset
 	// walletV  atomic.Value // *WalletState
 	balanceV atomic.Value // *WalletBalance
-
 }
 
 // makerBot is a *Core extension that enables operation of a market-maker bot.
-// Given an order for L lots, every epoch the makerBot will...
-//  1. Calculate a "basis price", which is based on DEX market data,
-//     optionally mixed (OracleWeight) with external market data.
-//  2. Calculate a "break-even spread". This is the spread at which tx fee
-//     losses exactly match profits.
-//  3. The break-even spread serves as a hard minimum, and is used to determine
-//     the target spread based on the specified gap strategy, giving the target
-//     buy and sell prices.
-//  4. Scan existing orders to determine if their prices are still valid,
-//     within DriftTolerance of the buy or sell price. If not, schedule them
-//     for cancellation.
-//  5. Calculate how many lots are needed to be ordered in order to meet the
-//     2 x L commitment. If low balance restricts the maintenance of L lots on
-//     one side, allow the difference in lots to be added to the opposite side.
-//  6. Place orders, cancels first, then buys and sells.
+// The strategy the makerBot follows depends on the engine.
 type makerBot struct {
 	*Core
```

Comment:
Will you remind me why bot creation and management is part of `Core` if (as I would expect) the bot itself wraps/manages a `Core` instance as here?  I'm not 100% against it, but I want to establish boundaries and guidelines for just how integrated with `Core` internals bots are supposed to get.

Reply:
I guess the reason is that the makerBot calls many of Core's functions. @buck54321 can probably explain more. I think the two options are doing it this way or doing what I've done with the `botEngine` input types.

Reply:
Ah, OK fair enough.  I guess I need to audit that to see.  Looks like it accesses unexported stuff like `Core.db`, `Core.cancelOrder`, `Core.notify`, `Core.syncBook`, `Core.ctx`,  etc.
I do wonder what it would take to have a bot function with `Core`'s exported API.  It certainly saves a lot of effort to be in the `client/core` package and expand `Core` rather than consume it as a black box.

Thanks for refactoring the bot and adding arb.  Good stuff!